docs/proposals/advanced-nginx-extensions.md

```diff
@@ -15,14 +15,1156 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+propagating into NGINX config. The Kubernetes API server will perform that validation when an Application developer
+creates a Custom resource of the CRD.
+
+#### SnippetsTemplate
+
+A template defined in SnippetsTemplate can be invalid (not follow go template or panic when executed). NGF must reject
+such templates and also not crash when executing them.
+
+A template can generate invalid NGINX configuration. When this happens, NGINX will continue to use the last valid
+configuration. However, any subsequent configuration updates (for example, caused by changes to an HTTPRoute) will not
+be possible until the invalid configuration is removed. Thus, a Cluster operator must carefully design the template.
+
+NGF will also validate the rest of SnippetsTemplate fields based on the restrictions mentioned in
+the [API section](#api-1).
+
+We will only allow one template per context because it will be easier to comprehend a SnippetsTemplate this way: compare
+one `http` template with multiple `http` templates scattered around in the spec.
+
+### Security Considerations
+
+As mentioned in the previous section, the Cluster operator should define validation rules to prevent invalid/malicious
+values from Custom resources created by Application developers from propagating into the NGINX config.
+
+However, it is possible that a template generates malicious values. Because of that, only the Cluster operator (a
+privileged user) should be able to create SnippetsTemplates. As a further precaution, we will disable SnippetsTemplate
+by default similarly to [SnippetsPolicy](#security-considerations).
+
+### Upgrades
+
+NGF inserts configuration generated from the templates of a SnippetsTemplate into various parts of the configuration
+that it already generates. As a result, the configuration generated from the templates must play well with
+the surrounding configuration, so that it doesn't break that configuration and vice versa.
+
+As we develop NGF, that surrounding configuration will expand (because we will implement more NGINX features) and also
+might change (for example, due to improvements). As a result, there is a risk that a SnippetsTemplate breaks after
+a Cluster operator upgrades NGF to the next version. Such risk shall be clearly documented in the SnippetsTemplate
+documentation.
+
+### Prior Arts
+
+- NGINX Ingress
+  Controller [Custom annotations](https://docs.nginx.com/nginx-ingress-controller/configuration/ingress-resources/custom-annotations/)
+- NGINX Instance
+  Manager [Config Templates](https://docs.nginx.com/nginx-management-suite/nim/about/templates/config-templates/)
+
+### Alternatives
+
+Instead of asking the Cluster operator to design a CRD, we can provide a ready container CRD. For example:
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsTemplate
+metadata:
+  name: rate-limiting-template
+spec:
+  kind: ClientSettingsPolicy
```

Comment:
Should this be `RateLimitingPolicy`? The template is configuring rate limiting...

Reply:
huh. replaced 7 (!) occurrences 🙄